File: backend/src/main/java/vn/chuongpl/badbook/features/product/Product.java
package vn.chuongpl.badbook.features.product;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;
import vn.chuongpl.badbook.common.enums.ProductCategory;
import vn.chuongpl.badbook.features.venue.Venue;

import java.math.BigDecimal;
import java.util.UUID;

@Entity @Table(name = "products")
@Getter @Setter @Builder @NoArgsConstructor @AllArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Product {

    @Id @GeneratedValue(strategy = GenerationType.UUID)
    @Column(columnDefinition = "uuid") UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "venue_id", nullable = false) Venue venue;

    @Column(nullable = false) String name;
    @Column(columnDefinition = "TEXT") String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default ProductCategory category = ProductCategory.OTHER;

    @Column(nullable = false, precision = 12, scale = 2) BigDecimal price;
    @Column(nullable = false) @Builder.Default String unit = "cái";
    @Column(nullable = false) @Builder.Default int stock = 0;
    @Column(name = "image_id") String imageId;
    @Column(name = "is_active") @Builder.Default boolean active = true;
}

File: apps/app-user/lib/models/product.dart
/// Mirrors backend `ProductResponse`.
class Product {
  final String id;
  final String venueId;
  final String name;
  final String? description;
  final String? category;
  final double price;
  final String? unit;
  final int stock;
  final String imageId;
  final bool active;

  Product({
    required this.id,
    required this.venueId,
    required this.name,
    this.description,
    this.category,
    required this.price,
    this.unit,
    required this.stock,
    required this.imageId,
    required this.active,
  });

  factory Product.fromJson(Map<String, dynamic> json) => Product(
        id: json['id'].toString(),
        venueId: json['venueId'].toString(),
        name: json['name'] as String? ?? '',
        description: json['description'] as String?,
        category: json['category'] as String?,
        price: (json['price'] as num?)?.toDouble() ?? 0.0,
        unit: json['unit'] as String?,
        stock: (json['stock'] as num?)?.toInt() ?? 0,
        imageId: json['imageId'] as String? ?? '',
        active: json['active'] as bool? ?? false,
      );
}
